src/hotspot/cpu/s390/interp_masm_s390.cpp

```diff
@@ -1005,19 +1005,21 @@ void InterpreterMacroAssembler::lock_object(Register monitor, Register object) {
 
   // markWord header = obj->mark().set_unlocked();
 
-  // Load markWord from object into header.
-  z_lg(header, hdr_offset, object);
-
   if (DiagnoseSyncOnValueBasedClasses != 0) {
     load_klass(tmp, object);
-    testbit(Address(tmp, Klass::access_flags_offset()), exact_log2(JVM_ACC_IS_VALUE_BASED_CLASS));
-    z_btrue(slow_case);
+    z_l(tmp, Address(tmp, Klass::access_flags_offset()));
+    assert((JVM_ACC_IS_VALUE_BASED_CLASS & 0xFFFF) == 0, "or change following instruction");
+    z_nilh(tmp, JVM_ACC_IS_VALUE_BASED_CLASS >> 16);
+    z_brne(slow_case);
```

Comment:
I see you implemented it in `compiler_fast_lock_lightweight_object` and changed it in `compiler_fast_lock_object` as well. Maybe that answers my question.

Reply:
Let's start from this one. I guess we made a mistake while pushing changes in https://github.com/openjdk/jdk/pull/18709. CC in case was being set to `true`. But we required it to be `EQ` or `NE`.  At that time the check which you have added were not there, to check the CC, and as a reason incorrect code was pushed :-( 

I wasn't sure whether I should open a separate PR to revert those change, Though I have opened the https://github.com/openjdk/jdk/pull/18987, but still I thought to ship it with this one.